Login signup & Common methods with utility & extensions

vineet-webonise opened the pull request.

saketPandhare-Webonise commented on SUNFLOWR/Podfile:
```diff
@@ -8,7 +8,19 @@ target 'SUNFLOWR' do
   pod 'Firebase/Analytics'
   pod 'GoogleSignIn'
   pod 'MoonKit'
-
```
I can see "JVFloatLabeledTextField" as well as "SkyFloatingLabelTextField" do we need two libraries for text field ?

vineet-webonise replied:
Removed "SkyFloatingLabelTextField"

saketPandhare-Webonise replied:
I can see "SkyFloatingLabelTextField" added in pod file

vineet-webonise replied:
Removed

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/ACProgressHUD/ACProgressHUD.swift:
```diff
@@ -0,0 +1,168 @@
+
+import UIKit
+
```
Use Cocoapod for "ACProgressHUD"   
pod 'ACProgressHUD-Swift', '~> 1.3'

vineet-webonise replied:
Yes we are using "ACProgressHUD" in Pod file, its dynamically considering this.

vineet-webonise replied:
This library is not updated to Swift 5.1

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/ACProgressHUD/ACHUD+Animation.swift:
```diff
@@ -0,0 +1,255 @@
+    case bounceFromTop
+    case none
+}
+
```
Try using pod for ACHUD instead of writing code

vineet-webonise replied:
This library is not updated to Swift 5.1, so files updated.

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Api/AlamofireUtilities.swift:
```diff
@@ -0,0 +1,339 @@
+import Alamofire
+
+class AlamofireUtilities {
+    
```
Write a single API Wrapper for all the GET/POST/PUT/DELETE calls .
Keeps Parameters Optionals that you thing are required in each one of them so that you can keep it nil for other calls 
For e.g if parameters are nil in GET call we can keep requestParameters as optionals  and pass as nil 
and pass actual parameters in POST call
Also pass HEADER parameters to the method as optionals .that will help moving forward

vineet-webonise replied:
Added wrapper

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Api/ApiConfigurationUtil.swift:
```diff
@@ -0,0 +1,74 @@
+     - returns: header contents in the form of dictionary.
+     */
+    static func getAuthHeaderContent() -> [String:String] {
+        
```
Instead of using "+" operator use string concatenation.

vineet-webonise replied:
Its one and same

saketPandhare-Webonise replied:
check the compile time here 
https://engineering.circle.com/swift-compiler-performance-tips-and-tricks-e86a53a5b42a

vineet-webonise replied:
Interesting, Done.

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Api/ConstantsApiKeys.swift:
```diff
@@ -0,0 +1,107 @@
+
+import Foundation
+
+extension Constants {
```
For constant use Struct instead of class as they are constant and we are not going to change it.

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Api/WebServiceUtil.swift:
```diff
@@ -0,0 +1,245 @@
+import UIKit
+
+class WebServiceUtil {
+    
```
Move this code to LoginViewModel. And try to Follow MVVM design pattern for ease of use

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Api/WebServiceUtil.swift:
```diff
@@ -0,0 +1,245 @@
+     */
+    func userLogin(requestJson:[String:AnyObject],
+                   success: @escaping ((_ userInfo:UserInfo)-> ()),
+                   failure: @escaping (_ errorMessage:String)-> ()) {
```
Create BASEURL and construct remaining URL out of it

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Common Methods/CommonMethods.swift:
```diff
@@ -0,0 +1,364 @@
+    }
+}
+
+//MARK:Persist userInfo 
```
We have initialised CoreData at the time of app creation and still using Realm here , which DB we are going to use ?Also remove the unwanted DB

vineet-webonise replied:
Done

vineet-webonise commented:
> Please review the PR once to remove the unwanted code.Things to check
> 1.Use Cocoapods where required
> 2.Finilize DB that we will be using CoreData/Realm and remove other
> 3.Remove unwanted code which is directly copied from other project keep only relevant part which is required here
> 4.Follow MVVM ,which will reduce complexity moving forward
> 5.Restructure API wrapper as commented

vineet-webonise commented:
> Please review the PR once to remove the unwanted code.Things to check
> 1.Use Cocoapods where required
> 2.Finilize DB that we will be using CoreData/Realm and remove other
> 3.Remove unwanted code which is directly copied from other project keep only relevant part which is required here
> 4.Follow MVVM ,which will reduce complexity moving forward
> 5.Restructure API wrapper as commented

Yes we are following MVVM in this project. API structure if to use Alamofire but in some case URLSession is also used for some limitation. Removed unwanted code. Removed unwanted DB for timebeing.

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Common Methods/CommonMethods.swift:
```diff
@@ -0,0 +1,364 @@
+    //TODO: Following code is for checking "Original images" in this build
+//    return prefix + Constants.Seperator.FORWARD_SLASH + Constants.Common.URL_INTERJECTION_ORIGIONAL + imageURL
+    
+    
```
Same here remove + and use string concatenation

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/Podfile.lock:
```diff
@@ -1,81 +1,98 @@
-    - AppAuth/ExternalUserAgent (= 1.3.0)
-  - AppAuth/Core (1.3.0)
-  - AppAuth/ExternalUserAgent (1.3.0)
-  - Firebase/Analytics (6.18.0):
```
Add Podlock to .gitignore

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Api/ApiConfigurationUtil.swift:
```diff
@@ -0,0 +1,71 @@
+        
+        return [HEADER_CONTENT_TYPE_KEY : CONTENT_TYPE_APPLICATION_JSON, HEADER_AUTH_KEY: auth]
+    }
+    
```
Give Proper name to the method and avoid name as "1", Also add method documentation

vineet-webonise replied:
removed & Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/SignUp/SignUpViewController.swift:
```diff
@@ -0,0 +1,187 @@
+    func addGradientToView() {
+        
+        viewWrapper.setNeedsLayout()
+        viewWrapper.layoutIfNeeded()
```
Remove unwanted code

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Login/LoginViewModel.swift:
```diff
@@ -0,0 +1,102 @@
+     */
+    func userLogin(requestJson:[String:AnyObject],
+                   success: @escaping ((_ userInfo:UserInfo)-> ()),
+                   failure: @escaping (_ errorMessage:String)-> ()) {
```
Move request URL in constant 
Use Alamofire Wrapper that we have created for API calling

vineet-webonise replied:
Added Wrapper

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Login/LoginViewController.swift:
```diff
@@ -0,0 +1,235 @@
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        setupUI()
+        //setup view model instance
```
Create ViewModel instance in ViewDidLoad method. As per ViewController lifecycle ,whenever you pop out a ViewController viewWillAppear will get called and a new instance will get created every time

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Home/HomeViewController.swift:
```diff
@@ -0,0 +1,152 @@
+        if error != nil{
+            print(error)
+            return
+        }
```
Move this method to viewModel and create separate method to handle operations after response received after API call

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Home/HomeViewController.swift:
```diff
@@ -0,0 +1,152 @@
+        GIDSignIn.sharedInstance()?.restorePreviousSignIn()
+    }
+    
+    
```
Create Separate extension in same file or Create Sperate Extension file to handle Delegate Methods.

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Home/HomeViewController.swift:
```diff
@@ -0,0 +1,152 @@
+        loginButton.permissions = ["public_profile", "email"]
+        loginButton.delegate = self
+    }
+    
```
Same comments as facebookLoginSetUp

vineet-webonise replied:
Done

saketPandhare-Webonise replied:
Use AutoLayout instead of hardcoding the stuff

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Home/HomeViewController.swift:
```diff
@@ -0,0 +1,152 @@
+        self.facebookLoginSetup()
+        self.googleLoginSetup()
+    }
+    
```
1.Add Method Documentation
2.First Create button and then add to subview 
3.Use AutoLayout to set the button frame ,avoid setting frame by providing hardcoded stuff 
If want to do it in code rather than UI ,set constraints through code

vineet-webonise replied:
Done

saketPandhare-Webonise replied:
Third Comment has not been resolved

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Common Methods/CommonMethods.swift:
```diff
@@ -0,0 +1,359 @@
+ */
+func labelSizeWithString(text: String, maxWidth : CGFloat,numberOfLines : Int) -> CGRect{
+    let label = UILabel(frame: CGRect(x:0, y:0,width: maxWidth,height: CGFloat.greatestFiniteMagnitude))
+    label.numberOfLines = numberOfLines
```
Remove blank spaces

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/AppDelegate.swift:
```diff
@@ -7,27 +7,45 @@ import CoreData
 class AppDelegate: UIResponder, UIApplicationDelegate {
     
-    static let googleClientId = "com.googleusercontent.apps.753427363835-h6jo4gaujfpohbt7kmk2150v7g396386"
+    var window: UIWindow?
```
remove blank spaces

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Api/WebServiceUtil.swift:
```diff
@@ -0,0 +1,186 @@
+            requestParameter: nil,
+            success: {
+                (response,statusCode) in
+                
```
Check error codes in the API wrapper in case of success .you can use range for checking 
e.g
```
if (status code is in range of 200...299) {
}
```

in the API wrapper class

vineet-webonise replied:
Removed

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Api/WebServiceUtil.swift:
```diff
@@ -0,0 +1,186 @@
+    ///   - failure: failure completion handler
+    func getLatestAppVersion(success : @escaping ((_ data:[String:AnyObject])-> ()),
+                             failure: @escaping (_ errorMessage:String)-> ()) {
+        
```
Try to use "\" instead of "+" operator wherever required

vineet-webonise replied:
Removed

vineet-webonise commented on SUNFLOWR/SUNFLOWR/Api/WebServiceUtil.swift:
```diff
@@ -0,0 +1,186 @@
+            requestParameter: deviceData,
+            success: {
+                (response,statusCode) in
+                //let statusCode = (response.response?.statusCode)!
```
Removed

vineet-webonise commented:
@saketPandhare-Webonise Can you please check the changes suggested. I have tried to make all the changes suggested along with Api wrapper class & removing unwanted code & libraries.

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Home/HomeExtension.swift:
```diff
@@ -0,0 +1,63 @@
+        if((AccessToken.current) != nil){
+            GraphRequest(graphPath: "me", parameters: ["fields": "id, name, first_name, last_name, email"]).start(completionHandler: { (connection, result, error) -> Void in
+                    if (error == nil){
+                        //parse the fields out of the result
```
Create separate method after response is received 
```
func callUserSignUp(response: response ) {
let fields = result as? [String:Any],
                             let firstName = fields["first_name"] as? String,
                             let lastName = fields["last_name"] as? String,
                             let email = fields["email"] as? String,
                             let id = fields["id"] as? String
                         {
                             self.viewModel.userSignUpApiCall(userEmail: email, userFirstName: firstName, userLastName: lastName,userPassword: Constants.Seperator.EMPTY_SPACE,dob: Constants.Seperator.EMPTY_SPACE,gender: Constants.Seperator.EMPTY_SPACE,experienceLevel: Constants.Seperator.EMPTY_SPACE, facebookUUID: id, googleUUID: Constants.Seperator.EMPTY_SPACE, fbAccessToken:AccessToken.current?.tokenString ?? Constants.Seperator.EMPTY_SPACE, googleAccessToken:Constants.Seperator.EMPTY_SPACE, deviceId: UUID().uuidString,    deviceToken: Constants.Seperator.EMPTY_SPACE, platform: Constants.ApiKeys.DeviceTokenApiKeys.IOS, platformVersion: "v1.0")
                         }
}
```

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Home/HomeViewController.swift:
```diff
@@ -0,0 +1,103 @@
+        // Automatically sign in the user.
+        GIDSignIn.sharedInstance()?.restorePreviousSignIn()
+    }
+    
```
Create Separate Extension for Delegate Methods

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Login/LoginViewController.swift:
```diff
@@ -0,0 +1,212 @@
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        setupUI()
+               
```
Create separate Methods for code in ViewController lifecycle methods 
like 
```
setDeviceTokenViewModel() -> This will have line no 45 ,46 
set textField() -> This will have line no 48,49
setSocialLogin() -> This will have line 52,53 
```
and so on

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Login/LoginViewController.swift:
```diff
@@ -0,0 +1,212 @@
+         if let token = AccessToken.current,
+               !token.isExpired {
+               // User is logged in, do work such as go to next view controller.
+           }
```
use Autolayout wherever required

vineet-webonise replied:
As of now we have set the width, leading, trailing & height constraints as per screen size which will act wisely.

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Login/LoginViewController.swift:
```diff
@@ -0,0 +1,212 @@
+        
+        switch sender.tag {
+        case ButtonTag.LOGIN.rawValue:
+            
```
Move below code to a method instead of folding the switch case 

```
func performActionOnLoginClick() {
guard IJReachability.isConnectedToNetwork() else {
                 showAlertWithMessage(message: Constants.AlertMessage.NO_INTERNET_CONNECTIVITY_MESSAGE, title: Constants.AlertMessage.NO_INTERNET_CONNECTIVITY_TITLE, cancelButtonTitle: Constants.AlertMessage.CANCEL_BUTTON_OK,
                                      viewController: self)
                 return
             }
             viewModel.performLogin(userEmailId: getTrimmedString(fieldText: textFieldEmailId.text!), userPassword:getTrimmedString(fieldText: textFieldPassword.text!))
}
```

And Call this method in switch case

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Login/LoginViewModel.swift:
```diff
@@ -0,0 +1,100 @@
+    func userLogin(userEmailId:String, userPassword: String, urlString: String,
+    success: @escaping ((_ userInfo:UserInfo)-> ()),
+    failure: @escaping (_ errorMessage:String)-> ()) {
+        
```
avoid  hardcoding of url ,use constant for baser and configure remaining urls out of it

vineet-webonise replied:
Its some how creating problem with guard let & lengthy use of constant string formation. Already tried.

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Login/LoginViewModel.swift:
```diff
@@ -0,0 +1,100 @@
+                                                    Constants.ApiKeys.LoginSignUpApiKeys.PLATFORM: Constants.ApiKeys.DeviceTokenApiKeys.IOS,
+
+                                                    Constants.ApiKeys.LoginSignUpApiKeys.DEVICE_TOKEN: Constants.Seperator.EMPTY_SPACE,
+                                                    Constants.ApiKeys.LoginSignUpApiKeys.PLATFORM_VERSION: Constants.ApiKeys.AppLogApiKeys.APP_VERSION]
```
Check for Blank spaces

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/SignUp/SignUpViewController.swift:
```diff
@@ -0,0 +1,146 @@
+    }
+    
+    //MARK:Buttonaction handlers
+    @IBAction func signUpButtonTapped(_ sender: UIButton) {
```
Move the Internet connection check code  to common class ,and just pass the view controller.to it 
```
showAlertWithMessage(message: Constants.AlertMessage.NO_INTERNET_CONNECTIVITY_MESSAGE, title: Constants.AlertMessage.NO_INTERNET_CONNECTIVITY_TITLE, cancelButtonTitle: Constants.AlertMessage.CANCEL_BUTTON_OK,
                                         viewController: self)
```
create method in common class and call that method by passing VC

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/SignUp/SignUpViewController.swift:
```diff
@@ -0,0 +1,146 @@
+    
+    @IBAction func buttonActionHandlers(sender: UIButton) {
+        switch sender.tag {
+        case ButtonTag.SIGNUP.rawValue:
```
Create Separate Method for Switch case instead of flooding the case

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/SignUp/SignUpViewModel.swift:
```diff
@@ -0,0 +1,176 @@
+    func userSignUp(requestJson:[String:String], urlString:String,
+                    success : @escaping ((_ userInfo:UserInfo)-> ()),
+                    failure: @escaping (_ errorMessage:String)-> ()) {
+       
```
Use Constant for URL

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Api/RestManager.swift:
```diff
@@ -0,0 +1,186 @@
+//
+//  RestManager.swift
+//  SUNFLOWR
+
```
Add Method documentations for methods in this class with parameter types

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/Common Methods/CommonMethods.swift:
```diff
@@ -0,0 +1,355 @@
+ - returns: Image URL
+ */
+func getImageByScreenScale(prefix:String, imageURL:String) -> String {
+    switch UIScreen.main.scale
```
Avoid using "+" operators

vineet-webonise replied:
Done

saketPandhare-Webonise commented on SUNFLOWR/Podfile:
```diff
@@ -8,7 +8,19 @@ target 'SUNFLOWR' do
   pod 'GoogleSignIn'
   pod 'MoonKit'
-
+  pod 'Alamofire'
```
Remove Alamofire If you are not using it

vineet-webonise replied:
Removed

saketPandhare-Webonise commented on SUNFLOWR/SUNFLOWR/SignUp/SignUpViewModel.swift:
```diff
@@ -0,0 +1,174 @@
+class SignUpViewModel: NSObject {
+    
+    var delegateSignUpViewModel: SignUpDelegate!
+    let rest = RestManager()
```
create baseURL 
`let BASE_URL  = "https://sunflowr-test.weboapps.com/api/v1"`

and then create url from it in constant

vineet-webonise replied:
Done
